lib/syslog/writer.go

```diff
@@ -243,11 +243,27 @@ func (c bufferedConn) Write(b []byte) (int, error) { return c.buf.Write(b) }
 
 func dialWriter(network string, address string, config *tls.Config) (w io.Writer, err error) {
 	var conn net.Conn
+	var dial func(string, string) (net.Conn, error)
 
 	if network == "tls" {
-		conn, err = tls.Dial("tcp", address, config)
+		network, dial = "tcp", func(network string, address string) (net.Conn, error) {
+			return tls.Dial(network, address, config)
+		}
 	} else {
-		conn, err = net.Dial(network, address)
+		dial = net.Dial
+	}
+
+	for attempt := 1; true; attempt++ {
+		if conn, err = dial(network, address); err == nil {
+			break
+		}
+
+		if attempt == 3 {
+			return
+		}
+
+		err = nil
+		time.Sleep(1 * time.Second)
```

Comment:
My only concern about this is that it could cause a log of "flush" operations to stack up, batched are flushed every 5 seconds, that could cause a potentially huge amount of connection retries and pending message batches in memory.

I haven't seen the issue take place over a long period of time, I think it's likely due to an ungraceful service restart on loggly's side that dropped the connection, retrying *right away* seems OK to me in this case.

Also with just 3 attempts, doing exponential backoff doesn't gain much, it's be retrying for 12 seconds instead of 3... if there's an actual outage going on it probably won't help much and just increase the pressure on ecs-logs's side.